# Project 3 -- Niharika Rai 

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [1]:
%%bash 
cd /anvil/projects/tdm/data/amazon/
pwd 

/anvil/projects/tdm/data/amazon


In [2]:
%%bash 
head /anvil/projects/tdm/data/amazon/tracks.csv

track_id|title|song_id|release|artist_id|artist_mbid|artist_name|duration|artist_familiarity|artist_hotttnesss|year|track_7digitalid|shs_perf|shs_work
TRMMMYQ128F932D901|"Silent Night"|SOQMMHC12AB0180CB8|"Monster Ballads X-Mas"|ARYZTJS1187B98C555|357ff05d-848a-44cf-b608-cb34b5701ae5|"Faster Pussy cat"|252.05506|0.649822100201|0.394031892714|2003|7032331|-1|0
TRMMMKD128F425225D|"Tanssi vaan"|SOVFVAK12A8C1350D9|"Karkuteillä"|ARMVN3U1187FB3A1EB|8d7ef530-a6fd-4f8f-b2e2-74aec765e0f9|Karkkiautomaatti|156.55138|0.439603966677|0.356992107756|1995|1514808|-1|0
TRMMMRX128F93187D9|"No One Could Ever"|SOGTUKN12AB017F4F1|Butter|ARGEKB01187FB50750|3d403d44-36ce-465c-ad43-ae877e65adc4|"Hudson Mohawke"|138.97098|0.643680572058|0.437503836595|2006|6945353|-1|0
TRMMMCH128F425532C|"Si Vos Querés"|SOBNYVR12A8C13558C|"De Culo"|ARNWYLR1187B9B2F9C|12be7648-7094-495f-90e6-df4189d68615|"Yerba Brava"|145.05751|0.448501159656|0.372349068517|2003|2168257|-1|0
TRMMMWA128F426B589|"Tangle Of Aspens"|SOHSBXH12A8C13B0

In [3]:
%%bash 
grep "Little Mix" /anvil/projects/tdm/data/amazon/tracks.csv

TRAJZHI128F932BB39|"I'm A Little Mixed Up"|SONDRHG12AB017F90A|"South Side Lady (Live in Netherlands 1973) (Blues Reference)"|ARWF1971187B9B66CC|9ebffe22-c6e0-49bc-a6c2-f79a97bc38b3|"Koko Taylor"|222.04036|0.621432272125|0.423620810234|1969|6069564|-1|0
TROJQKV128F92D6730|"I'm A Little Mixed Up"|SOTNQQN12AAF3B2B46|"Rhythm & Blues Vol. 8"|ARJ0FXD1187B99F639|8b84d752-9f4e-498e-ad27-23a5de2b69c5|"Betty James"|165.27628|0.377304320601|0.203963938841|1993|3628911|-1|0


In [4]:
%%bash 
grep "Waiting for Blue" /anvil/projects/tdm/data/amazon/tracks.csv

TRQOXRL128F930D94A|"Waiting for Blue"|SOHFYOF12AB017FA1A|"Testing Superstition"|ARZGSBD1187B9919F2|2be0a553-3f67-4ee8-84d3-0824d9889006|"The June Spirit"|245.49832|0.481722049976|0.306129858047|2002|5459913|-1|0


In [1]:
%%bash 
grep "Waiting for Blue" /anvil/projects/tdm/data/amazon/tracks.csv | cut -d'|' -f7

"The June Spirit"


In [6]:
%%bash
grep -c "Michael Jackson" /anvil/projects/tdm/data/amazon/tracks.csv

253


In [7]:
%%bash
grep -i -c "X Marks the Pedwalk" /anvil/projects/tdm/data/amazon/tracks.csv

47


In [8]:
%%bash
grep -i -n "Rainbow in the Night" /anvil/projects/tdm/data/amazon/tracks.csv

351687:TRZHTHL128F42AB7ED|"Rainbow In The Night"|SOUOOBS12A8C13FA6C|"You Are Being Lied To"|ARQMX4G1187FB5934E|90f9fc57-7b6f-4229-a69f-84d2644882ff|Kinesis|214.80444|0.466848408257|0.360992335286|2005|3460954|-1|0


In [9]:
%%bash
grep -i "garlic" /anvil/projects/tdm/data/amazon/tracks.csv

TRHXDRN128F429F4BB|"Garlic Blues"|SONFGXK12A8C13FA75|"Sylvester Weaver Vol. 2 (1927)"|ARPJNT81187B98AD36|1e1fa4a5-3bd6-4a1c-9b1d-5c613a741500|"Sylvester Weaver"|180.47955|0.370703603519|0.283196596109|0|1871681|-1|0
TRCCKZJ128F92C4F7B|"Have Got Garlic Hanging On My Front Door"|SOYAHQF12A8C140C04|Signals|ARTUHAL1187B98B477|b93197b9-c613-4ecc-8146-4f7a11407f70|Marasma|400.79628|0.422275305174|0.215381650322|1998|3545788|-1|0
TRFQZIJ128F92FEEDB|"King Garlic"|SOWHCIV12AB0181A45|Spaceguerilla|AR5YT6E119B340313C|dccff302-f95e-46ce-8f95-a8ddd811b5d9|"Missus Beastly"|396.95628|0.389691318086|0.297792838565|0|4407167|-1|0
TRIYIKE128F423BB77|"Boeuf gros sel [Instrumental]"|SOHIEQA12A8AE46D0C|Garlick|ARUZB9Q1187FB45C9C|9ec13aa9-43e9-4a86-95be-d2b5c188dd20|"Hugues Aufray"|154.14812|0.561993937673|0.363323041772|1972|1401763|-1|0
TRTLHNN128F93273F6|"Garlic Bread"|SOSWOZF12AF72A3F30|"ASBO 4 Life"|ARJXVSV1187B9A252F|5167695e-3a2a-4311-9990-b7e2d13686a2|"Goldie Lookin Chain"|226.21995|0.640784077463|0

In [10]:
%%bash
grep -i -w "garlic" /anvil/projects/tdm/data/amazon/tracks.csv

TRHXDRN128F429F4BB|"Garlic Blues"|SONFGXK12A8C13FA75|"Sylvester Weaver Vol. 2 (1927)"|ARPJNT81187B98AD36|1e1fa4a5-3bd6-4a1c-9b1d-5c613a741500|"Sylvester Weaver"|180.47955|0.370703603519|0.283196596109|0|1871681|-1|0
TRCCKZJ128F92C4F7B|"Have Got Garlic Hanging On My Front Door"|SOYAHQF12A8C140C04|Signals|ARTUHAL1187B98B477|b93197b9-c613-4ecc-8146-4f7a11407f70|Marasma|400.79628|0.422275305174|0.215381650322|1998|3545788|-1|0
TRFQZIJ128F92FEEDB|"King Garlic"|SOWHCIV12AB0181A45|Spaceguerilla|AR5YT6E119B340313C|dccff302-f95e-46ce-8f95-a8ddd811b5d9|"Missus Beastly"|396.95628|0.389691318086|0.297792838565|0|4407167|-1|0
TRTLHNN128F93273F6|"Garlic Bread"|SOSWOZF12AF72A3F30|"ASBO 4 Life"|ARJXVSV1187B9A252F|5167695e-3a2a-4311-9990-b7e2d13686a2|"Goldie Lookin Chain"|226.21995|0.640784077463|0.462737056844|2009|4609633|-1|0
TRSAEGK128E0788EA2|Garlic|SOSIPTC12A6701C549|"Disappear Here"|ARX13A01187B99FF60|c76e4d1c-06ff-4594-b31c-4a93d9c1cb22|"Silver Sun"|188.47302|0.514594286426|0.362416410941|0|108

In [11]:
%%bash
grep -i "garlic" /anvil/projects/tdm/data/amazon/tracks.csv | grep -ivw "garlic" | wc -l

3


1.3 The June Spirit sang "Waiting for Blue". I located the artist's name when running head and found that it is in the 7th column. I found that the separator is "|" instead of "," so I used that accordingly to locate the artist's name. 
1.4 A total of 253 lines contain "Michael Jackson". 
1.5 A total of 47 lines contain "X Marks the Pedwalk". 
1.6 The line in the file that contains "Rainbow in the Night" is 351687. 
1.9 A total of 3 lines contain "garlic" as a substring but not as a whole word. I used ChatGPT to figure out the code structure to achieve this. The grep -ivw "garlic" excludes the lines where "garlic" appears as a whole word because -v inverts the match. The wc -l counts the remaining lines and gives us the answer. 

## Question 2

In [1]:
%%bash 
cd /anvil/projects/tdm/data/amazon/
pwd 

/anvil/projects/tdm/data/amazon


In [2]:
%%bash 
grep -i -c "apple" /anvil/projects/tdm/data/amazon/*.csv

/anvil/projects/tdm/data/amazon/Reviews.csv:10635
/anvil/projects/tdm/data/amazon/amazon_fine_food_reviews.csv:10635
/anvil/projects/tdm/data/amazon/tracks.csv:1028


In [3]:
%%bash 
grep -r "The Factory Girl" /anvil/projects/tdm/data/movies_and_tv/*

grep: /anvil/projects/tdm/data/movies_and_tv/imdb.db: binary file matches


/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0004888	1	The Factory Girl	US	\N	working	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt1132235	1	The Factory Girl	FR	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt1132235	2	The Factory Girl	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt2007481	1	The Factory Girl	US	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt2007481	2	The Factory Girl	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt2104110	1	The Factory Girl's Honour	GB	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt2104110	2	The Factory Girl's Honour	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt3182596	10	The Factory Girl	XWW	en	\N	alternative spelling	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.principals

grep: /anvil/projects/tdm/data/movies_and_tv/imdb2024.db: binary file matches


In [4]:
%%bash 
grep -i -r "pointer" /anvil/projects/tdm/data/movies_and_tv/*

grep: /anvil/projects/tdm/data/movies_and_tv/imdb.db: binary file matches


/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0006451	1	Bobby Bumps and His Pointer Pup	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0006451	2	Bobby Bumps and His Pointer Pup	US	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0017829	11	The West Pointer	US	\N	working	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	10	El pointer	ES	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	13	The Pointer	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	2	Como Treinar Um Pointer	BR	\N	\N	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	4	The Pointer	US	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	8	El pointer	AR	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0146003	1	Pointe

grep: /anvil/projects/tdm/data/movies_and_tv/imdb2024.db: binary file matches


/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes/rotten_tomatoes_critic_reviews.csv:m/1112954-rookie,Chuck Wilson,False,L.A. Weekly,Rotten,,2002-06-01,"This is a boy's movie, a fact director John Lee Hancock ... and screenwriter Mike Rich ... make clear with such images as an anonymous boy and his dad in the stands watching the big game together, the father whispering pointers to his son."
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes/rotten_tomatoes_critic_reviews.csv:m/1155467-glory_road,Matt Pais,False,Metromix.com,Fresh,3/4,2006-05-04,"A darn good story, told with enough conviction and impact to connect with people who don't know the difference between a three-pointer and a field goal."
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes/rotten_tomatoes_critic_reviews.csv:m/1201402-17_again,Carrie Rickey,True,Philadelphia Inquirer,Fresh,3/4,2009-04-16,"For me, 17 Again is mildly diverting. For the tweens who are its target audience, Efron's outside shot is a three

2.2 The number of lines in each of the csv files that contain "apple" and are case-insensitive are written below. 
Reviews.csv:10635
amazon_fine_food_reviews.csv:10635
tracks.csv:1028

2.3 The option that can make the command work is "-r" because it tells grep to search all files recursively inside the directories. I used ChatGPT to learn about this command. 

## Question 3

In [4]:
%%bash 
grep "Michael Jackson" /anvil/projects/tdm/data/amazon/tracks.csv | wc -l

253


In [6]:
%%bash 
grep -i "X Marks the Pedwalk" /anvil/projects/tdm/data/amazon/tracks.csv | wc -l

47


In [11]:
%%bash 
head /anvil/projects/tdm/data/amazon/Reviews.csv

Id,ProductId,UserId,ProfileName,HelpfulnessNumerator,HelpfulnessDenominator,Score,Time,Summary,Text
1,B001E4KFG0,A3SGXH7AUHU8GW,delmartian,1,1,5,1303862400,Good Quality Dog Food,I have bought several of the Vitality canned dog food products and have found them all to be of good quality. The product looks more like a stew than a processed meat and it smells better. My Labrador is finicky and she appreciates this product better than  most.
2,B00813GRG4,A1D87F6ZCVE5NK,dll pa,0,0,1,1346976000,Not as Advertised,"Product arrived labeled as Jumbo Salted Peanuts...the peanuts were actually small sized unsalted. Not sure if this was an error or if the vendor intended to represent the product as ""Jumbo""."
3,B000LQOCH0,ABXLMWJIXXAIN,"Natalia Corres ""Natalia Corres""",1,1,4,1219017600,"""Delight"" says it all","This is a confection that has been around a few centuries.  It is a light, pillowy citrus gelatin with nuts - in this case Filberts. And it is cut into tiny squares and then liberally co

In [9]:
%%bash 
cat /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d, -f4,7,9 | head

ProfileName,Score,Summary
delmartian,5,Good Quality Dog Food
dll pa,1,Not as Advertised
"Natalia Corres ""Natalia Corres""",4,"""Delight"" says it all"
Karl,2,Cough Medicine
"Michael D. Bigham ""M. Wassir""",5,Great taffy
Twoapennything,4,Nice Taffy
David C. Sullivan,5,Great!  Just as good as the expensive brands!
Pamela G. Williams,5,"Wonderful
R. James,5,Yay Barley


In [10]:
%%bash 
head /anvil/projects/tdm/data/amazon/music.txt | tr ' ' '_'

"I_love_this_CD.__So_inspiring!"
"Love_it!!__Great_seller!"
"I_bought_this_on_cassette_tape_in_the_80's._So_inspirational_to_me_back_then._Came_across_it_again_and_needed_to_be_uplifted!_Keith_Green's_music_still_works_magic._I_bought_CD's_for_my_girls_to_play_in_their_cars"
"as_good_as_i_remember_back_when_i_bought_the_one_in_the_80's"
"One_of_my_very_favourite_albums_from_one_of_my_very_favourite_singers.__I_was_happy_to_see_I_could_replace_the_old_worn_cassettes_from_years_ago."
"THank_you_Jesus_Lord_God,_that_brother_Green's_music_is_still_sounding_though_he_is_Home_with_you_now."
"I_recall_loving_his_other_albums_and_maybe_this_one_too,_forgot_but_I_figured_on_some_of_these_artists_seems_like_one_good_album_and_all_good_albums..especially_in_Christian_music..seemed_when_they_got_into_it_they_stayed_into_it_and_so_good_to_double_check_though_if_want_too_but_it_is_a_possible_very_good_album_because_I_usually_recall_if_too_bad_of_one_and_I_don't_on_this_one"
"Keith_Green's_music_are_

In [16]:
%%bash 
cat /anvil/projects/tdm/data/amazon/Reviews.csv | head -n 76 | grep -i "chocolate" | cut -d, -f2 

B00144C10S
B004N5KULM
B004N5KULM
B004N5KULM
B000E7VI7S
B0059WXJKM


3.1 A total of 253 lines contain "Michael Jackson". I used the 'wc -l' command after the "|" symbol to find the line count. 
3.2 A total of 47 lines contain "X Marks the Pedwalk". 
3.3 I first printed the first few lines and located the position of the ProfileName, Score, and Summary columns. I used the same code structure from Dr. Ward's video to formulate my code. 
3.4 I used ChatGPT to learn how to use the 'tr' command after I inserted the pipeline "|" symbol. 
3.5 I followed Dr. Ward's video and sample code to extract the ProductId based on the 3 given conditions and used the pipeline "|" symbol 3 times. 

## Question 4

In [17]:
%%bash 
cat /anvil/projects/tdm/data/amazon/Reviews.csv | head -n 76 | grep -i "chocolate" | cut -d, -f2 | sort | uniq

B000E7VI7S
B00144C10S
B004N5KULM
B0059WXJKM


In [18]:
%%bash 
cat /anvil/projects/tdm/data/amazon/Reviews.csv | head -n 76 | grep -i "chocolate" | cut -d, -f2 | sort | uniq -c

      1 B000E7VI7S
      1 B00144C10S
      3 B004N5KULM
      1 B0059WXJKM


In [24]:
%%bash 
cat /anvil/projects/tdm/data/amazon/Reviews.csv | head -n 76 | grep -i "chocolate" | cut -d, -f2 | sort | uniq -c | sort -nr

      3 B004N5KULM
      1 B0059WXJKM
      1 B00144C10S
      1 B000E7VI7S


In [23]:
%%bash 
cat /anvil/projects/tdm/data/amazon/Reviews.csv | grep -i "chocolate" | cut -d, -f2 | sort | uniq -c | sort -nr | head -n 10

    259 B002NHYQAS
    228 B004U49QU2
    201 B001OCKIP0
    195 B001EQ55RW
    190 B004U43ZO0
    181 B0076MLL12
    181 B005K4Q68Q
    181 B005K4Q4KG
    181 B005K4Q1VI
    181 B005K4Q1T0


For the above code output, I used the sample code in Dr. Ward's video as reference. To sort numerically in descending order, I used "-nr" instead of "-n". The "r" indicates that the reverse order be applied when sorting the counts. 

## Question 5

In [53]:
%%bash 
cd /anvil/projects/tdm/data/amazon/ 
pwd 

/anvil/projects/tdm/data/amazon


In [12]:
%%bash 
grep -i -E "love|hate" /anvil/projects/tdm/data/amazon/Reviews.csv | wc -l

186028


In [55]:
%%bash 
grep "^[A-Z]" /anvil/projects/tdm/data/amazon/Reviews.csv

Id,ProductId,UserId,ProfileName,HelpfulnessNumerator,HelpfulnessDenominator,Score,Time,Summary,Text


In [8]:
%%bash 
head /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d, -f9 | grep "^[A-Z]"

Summary
Good Quality Dog Food
Not as Advertised
Cough Medicine
Great taffy
Nice Taffy
Great!  Just as good as the expensive brands!
Yay Barley


In [6]:
%%bash 
cut -d, -f9 /anvil/projects/tdm/data/amazon/Reviews.csv | grep "^[A-Z]" | wc -l

439383


In [9]:
%%bash 
cut -d, -f9 /anvil/projects/tdm/data/amazon/Reviews.csv | grep -i "great$" | wc -l

4807


In [10]:
%%bash 
cut -d, -f10 /anvil/projects/tdm/data/amazon/Reviews.csv | grep -P "\d.*\d" | wc -l

60831


Since the filename was not specified, I used the Reviews.csv since it has been continuously used for the past few questions. Without the filename, I was getting an error. 

5.2 A total of 186028 lines contain either "love" or "hate" in the Reviews.csv file. I used "-i" since the question specifies case-insensitive. I used ChatGPT to learn how to implement a regular expression through code.   
5.3 The lines that begin with capital letters were only the column names. I had to use ChatGPT to find out how to use the "^" symbol with my code. This symbol means the start of the line.  
5.4 I knew that the summary column was in the 9th position and adjusted the code for the previous sub-question to only the first few rows using head. 
5.5 A total of 439383 lines in the Summary column of the whole file begin with a capital letter. I replaced 'head' with 'cut -d, -f9' since that also reads the whole file and counted the lines using pipeline and the 'wc -l' command. 
5.6 A total of 4807 lines in the summary column end with "great". I used ChatGPT to understand where to position the "$" symbol. This symbol indicates the end of the line.  
5.7 A total of 60831 reviews contain at least 2-digit numbers. I knew that the text column was in the 10th position, and I used the '-P "\d.*\d"' next to my grep command to ensure that the matching lines contained at least 2-digit numbers. I learnt about this code syntax from ChatGPT. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
